### PR TITLE
Add support for Doctrine DBAL 4 and ORM 3

### LINKED ISSUE #1774: Doctrine ORM 3 support
Hi folks!

You may have noticed a couple beta releases of `doctrine/orm` :tada: 

https://github.com/doctrine/orm/releases/tag/3.0.0-beta2

We'd love to get feedback from you on this, please let us know if you encounter any difficulties integrating this with Sonata.

### PR DESCRIPTION
**Pull request by dmaicher**
## Subject

I am targeting this branch, because its BC.

Fixes https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/1774

## Changelog

```markdown
### Added
- compatibility with `doctrine/dbal` 4 and `doctrine/orm` 3
```

### CODE REVIEW

**Comment by dmaicher** [comment]
~~I'm blocked by one test fail where I'm not sure if its an ORM bug or an issue on our side :confused:~~

```
[critical] Uncaught PHP Exception PHPUnit\Framework\Error\Warning: "Undefined property: Doctrine\ORM\Mapping\OneToOneInverseSideMapping::$joinColumns" at IdentityFunction.php line 40
``` 
See https://github.com/doctrine/orm/blob/3907872046b9a697860f51266714c2eb8e3b19d0/src/Query/AST/Functions/IdentityFunction.php#L40

~~Looks more like an ORM issue to me. @greg0ire any idea? Shall I open an issue for it on ORM?~~

@greg0ire never mind. It was an issue on our end :+1: found it after some more debugging. Sorry for the noise.

**Comment by dmaicher** [comment]
Tests are finally green :smile: 

Shall we move ahead without https://github.com/sonata-project/EntityAuditBundle/issues/59 ? Shall I maybe create a separate workflow file (not managed by devkit) for the time being so we are running CI on DBAL 4 and ORM 3?

File: tests/Model/ModelManagerTest.php [hunk 1]
```diff
@@ -643,35 +642,35 @@ public function testFailingBatchDelete(string $expectedExceptionMessage, ?array
             ->method('getConnection')
             ->willReturn($connection);
 
-        $hydrator = $this->createMock(SimpleObjectHydrator::class);
-        $hydrator
+        $query = $this->createMock(Query::class);
+        $query
             ->expects(static::once())
             ->method('toIterable')
```
Comment (dmaicher) [inline tests/Model/ModelManagerTest.php:648]: this is final on `AbstractHydrator` now. So we cannot mock it anymore.

**Comment by VincentLanglet** [comment]
> Tests are finally green 😄
> 
> Shall we move ahead without [sonata-project/EntityAuditBundle#59](https://github.com/sonata-project/EntityAuditBundle/issues/59) ? Shall I maybe create a separate workflow file (not managed by devkit) for the time being so we are running CI on DBAL 4 and ORM 3?

I'm ok with this strategy

**Review by VincentLanglet — APPROVED** [review]

**Review by core23 — APPROVED** [review]

**Pull request merged**